### ISSUE TITLE
Repetition penalty is gauge-dependent. Proposing opt-in normalized variant.

### ISSUE THREAD

**Issue by captainpete** [huggingface/transformers#47595]
### Feature request

Add an opt-in mode that applies `repetition_penalty` to normalized log-probs instead of raw logits.
Default behavior unchanged.

### Motivation

The rep penalty currently branches on the sign of each raw logit.
This was added in 2019 as a workaround (#2302, #2303) for models with different logit zero-points.

The branching workaround is a departure from the original CTRL paper. #2303 fixed the negative-logit case but introduced a "kink" that causes problems. In practice, this kink negatively impacts output quality when greedy sampling.

See [arXiv:2607.09791](https://arxiv.org/abs/2607.09791) for the write-up. Tested across 5 models (up to 7B). On JSONSchemaBench the valid outputs drop from 97% to 23% at θ=1.3. Repo for the paper lives at [captainpete/repetition-penalty-gauge](https://github.com/captainpete/repetition-penalty-gauge).

After running the experiments I've authored a fix candidate (apply the penalty to log-probs, always <= 0, so we can remove the branch). It composes cleanly with the rest of the pipeline. In my testing it doesn't see the same quality drop. The beam sampler already takes this approach ([beam uses `log_softmax`](https://github.com/huggingface/transformers/blob/44752c8dd99f3fb0da23006dc4fde4a07d9c417f/src/transformers/generation/utils.py#L3369-L3372)).

I'm thinking that if this fix is adopted it should be opt-in plus some docs.

#### The TL;DR for the arXiv note

The current penalty branches like so:

```python
score = torch.where(score<0, score*self.penalty, score/self.penalty)
```

The zero-point of a snapshot's logits is arbitrary and bridges zero in practice.
There's just nothing in the training loop that enforces logits to play nice here. Softmax is shift-invariant, so they end up all over the place.

This means a rep penalty of 1.3 means something totally different between snapshots and across models.
It also means uneven handling of penalties within a snapshot that bridges the zero-point.

The branching workaround runs through the inference ecosystem for this kind of rep penalty, but HF is at the root of that lineage. Figured I'd discuss here with you folk first and get your take.

#### Triage snippet

```python
import torch
from transformers import RepetitionPenaltyLogitsProcessor

input_ids = torch.tensor([[0, 1]])
scores = torch.arange(10, dtype=torch.float).unsqueeze(0) / 10 - 0.5 # bridges zero

penalty_proc = RepetitionPenaltyLogitsProcessor(penalty=1.3)
p1 = torch.softmax(penalty_proc(input_ids, scores), dim=-1)
p2 = torch.softmax(penalty_proc(input_ids, scores + 100.0), dim=-1)  # should be no-op under softmax
print(torch.allclose(p1, p2))  # False
```

#### Related issues

#2302, #2303, #26902, #36642, #37625
Note: @tysam-code pointed out this same issue with the math in ggml-org/llama.cpp#2970 in 2023.

### Your contribution

I have a working implementation on a fork [captainpete:transformers:normalized-repetition-penalty](https://github.com/huggingface/transformers/compare/main...captainpete:transformers:normalized-repetition-penalty) that adds:

- `normalize` argument on `RepetitionPenaltyLogitsProcessor` and `EncoderRepetitionPenaltyLogitsProcessor` (opt-in)
- Exposed as `repetition_penalty_normalize` and `encoder_repetition_penalty_normalize` in `GenerationConfig`. (opt-in)
- A few tests
- I tried to keep this surgical, around +50/-10 in the hot path.

I'll open a PR if there's appetite for this. From what I've seen there are related issues in other inference stacks, but most of the implementations deferred to the workaround in #2303.

If we don't end up going for the normalize approach then maybe a docs note can steer people toward other penalty styles.

**Comment by clanzhang** [comment]
Hi @captainpete, thanks for the thorough write-up and the accompanying paper — the gauge-dependence argument is convincing, especially the JSONSchemaBench results at θ=1.3.

I'd like to help get this into transformers. A few questions before I start a PR:

1. **Integration point** — are you thinking of adding a new flag to `RepetitionPenaltyLogitsProcessor` (e.g. `normalize=True`), or a separate processor class entirely? A flag keeps the API surface small, but a new class avoids any risk to the existing codepath.

2. **GenerationConfig wiring** — I assume this would need a new field in `GenerationConfig` (something like `repetition_penalty_normalized: bool = False`) so it can be set via `model.generate(...)` and serialized. Does that match your expectation?

3. **Beam search parity** — you mentioned beam already uses `log_softmax`. Should the normalized path simply reuse that logic, or keep a distinct implementation for clarity?

I'm happy to take a first pass at the PR if you're open to collaborating. I can work from your repo (`repetition-penalty-gauge`) as the reference implementation and adapt it to the transformers codebase conventions. Would you be available to review?

cc @gante @molbap (generation team, for visibility on whether this is in scope)

**Comment by captainpete** [comment]
Thanks @clanzhang, I have a working branch (linked in the issue), it covers what you've asked.
I will open a PR myself pending feedback from the generation team.

**Comment by qgallouedec** [comment]
Hey @captainpete, thanks for the detailed write-up and for keeping the change opt-in. The branch looks nicely scoped, so please go ahead and open the PR you mentioned, that's the fastest way to get a real design review. We could go from there if it's worth adding

One routing note: the current generate owners on the who-can-help list are @zucchini-nlp and @Cyrilvallez, not the folks cc'd above, so tagging them on the PR itself should get you a faster response.

Leaving this open as a feature request in the meantime.

**Comment by captainpete** [comment]
Update: the branch is rebased and pushed. PR open (#48642). Tagged zucchini-nlp and Cyrilvallez as suggested.

Note: I released an updated arXiv (2607.09791v2) with a matched-suppression comparison and a per-model calibration map in the meantime. The short version: the normalized form is gauge-invariant by construction and statistically no worse on any quality metric at matched suppression, but the same `repetition_penalty` value is much weaker on log-probabilities. That's about as expected, but it was interesting to match values across different models to see just how different the original raw value was in effect when mapped to a normalized value.

Anyway, keen to get feedback on the PR. @qgallouedec thanks for inviting the contribution.
